File: src/main/java/org/chousik/collection/builder/collectors/EnumCollector.java
package org.chousik.collection.builder.collectors;

import org.chousik.collection.Color;
import org.chousik.collection.validators.IValidator;
import org.chousik.exception.InvalidDataError;
import org.chousik.exception.ScriptExecutionError;
import org.chousik.handlers.RunHandler;

import java.util.NoSuchElementException;
import java.util.Scanner;
import java.lang.Enum;
import java.util.function.Function;
import java.util.function.Supplier;

public abstract class EnumCollector<T extends Enum> implements ICollector<T, String> {
    private final Boolean isScript;
    private final Scanner scanner;

    public EnumCollector() {
        this.isScript = RunHandler.mode();
        this.scanner = RunHandler.getMainScaner();
    }

    protected T askEnum(String name, IValidator<String> validator, Function<String, T> method1, Supplier<String> method2) throws ScriptExecutionError {
        while (true) {
            try {
                if (!isScript) {
                    System.out.println("Доступные варианты поля " + name + ": " + method2.get());
                    System.out.println("Выберите один из вариантов");
                    System.out.print(System.getProperty("user.name")+"> ");
                }
                String strValue = scanner.nextLine().trim().toUpperCase();
                if (strValue.isEmpty()) {
                    strValue = null;
                }
                validator.valide(strValue);
                return method1.apply(strValue);
            } catch (InvalidDataError e) {
                if (isScript) {
                    throw new ScriptExecutionError("Введено неккоретное значение из списка.");
                }
                System.out.println("Введено неккоретное значение из списка.");
            } catch (NullPointerException e) {
                if (isScript) {
                    throw new ScriptExecutionError("Значение не может быть null");
                }
                System.out.println("Значение не может быть null");
            } catch (NoSuchElementException e) {
                if (isScript) {
                    throw new ScriptExecutionError("Ошибка во время ввода данных коллекции из файла. Конец файла.");
                }
                System.out.println("Не нажимай Ctrl+D((((");
                System.exit(0);
            } catch (Exception e) {
                if (isScript) {
                    throw new ScriptExecutionError("Непридвиденная ошибка");
                }
                System.out.println("Непридвиденная ошибка");
                System.exit(0);
            }
        }
    }
}

File: src/main/kotlin/collection/builder/collectors/EnumCollector.kt
package org.chousik.collection.builder.collectors

import org.chousik.collection.validators.IValidator
import exeption.InvalidDataError
import exeption.ScriptExecutionError
import org.chousik.handlers.RunHandler
import java.util.*
import java.util.function.Function
import kotlin.system.exitProcess

abstract class EnumCollector<T : Enum<*>?> : ICollector<T, String?> {
    private val isScript = RunHandler.mode()
    private val scanner: Scanner = RunHandler.getMainScanner()

    protected fun askEnum(
        name: String,
        validator: IValidator<String?>,
        method1: Function<String, T>,
        value: String
    ): T {
        while (true) {
            try {
                if (!isScript) {
                    println("Доступные варианты поля $name: $value")
                    println("Выберите один из вариантов")
                    print(System.getProperty("user.name") + "> ")
                }
                var strValue: String? = scanner.nextLine().trim { it <= ' ' }.uppercase(Locale.getDefault())
                if (strValue!!.isEmpty()) {
                    strValue = null
                }
                validator.valide(strValue)
                return method1.apply(strValue!!)
            } catch (e: InvalidDataError) {
                if (isScript) {
                    throw ScriptExecutionError("Введено некоренное значение из списка.")
                }
                println("Введено некоренное значение из списка.")
            } catch (e: NullPointerException) {
                if (isScript) {
                    throw ScriptExecutionError("Значение не может быть null")
                }
                println("Значение не может быть null")
            } catch (e: NoSuchElementException) {
                if (isScript) {
                    throw ScriptExecutionError("Ошибка во время ввода данных коллекции из файла. Конец файла.")
                }
                println("Не нажимай Ctrl+D((((")
                exitProcess(0)
            } catch (e: Exception) {
                if (isScript) {
                    throw ScriptExecutionError("Непредвиденная ошибка")
                }
                println("Непредвиденная ошибка")
                exitProcess(0)
            }
        }
    }
}